processing/src/main/java/org/apache/druid/query/groupby/GroupByResourcesReservationPool.java

```diff
@@ -125,17 +146,22 @@ public void reserve(QueryResourceId queryResourceId, GroupByQuery groupByQuery,
   @Nullable
   public GroupByQueryResources fetch(QueryResourceId queryResourceId)
   {
-    return pool.get(queryResourceId);
+    GroupByQueryResources resource = pool.get(queryResourceId).get();
+    assert resource != null;
+    return resource;
   }
 
   /**
    * Removes the entry corresponding to the unique id from the map, and cleans up the resources.
    */
   public void clean(QueryResourceId queryResourceId)
   {
-    GroupByQueryResources resources = pool.remove(queryResourceId);
-    if (resources != null) {
-      resources.close();
+    AtomicReference<GroupByQueryResources> resourcesReference = pool.remove(queryResourceId);
+    if (resourcesReference != null) {
+      GroupByQueryResources resource = resourcesReference.get();
+      // Reference should refer to a non-empty resource
+      assert resource != null;
```

Comment:
Better to use `DruidException.defensive` than `assert`.

Reply:
Since all the handling was done in a single class, I figured it would be fine to have an `assert` statement. Modified it to DruidException.defensive().